dvc/system.py

```diff
@@ -28,6 +35,8 @@ def hardlink(source, link_name):
         except OSError as exc:
             raise DvcException("failed to link") from exc
 
+        _verify_link(link_name, System.is_hardlink, "hardlink")
```

Comment:
A bit worried about a potential performance hit, so might just add these checks to `dvc version`. Or maybe to cache type caching mechanism that we have, so we could only run it once. Will take a closer look soon.